torch.fft: Two dimensional FFT functions

peterbell10 opened the pull request:
This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45164 torch.fft: Two dimensional FFT functions**

Differential Revision: [D24363639](https://our.internmc.facebook.com/intern/diff/D24363639)

peterbell10 committed:
torch.fft: Two dimensional FFT functions

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7787,6 +7787,23 @@
   use_c10_dispatcher: full
   variants: function
 
+# Would like this to be "int[1]? dim=[-2,-1]" but this fails with an error at import time
+- func: fft_fft2(Tensor self, int[1]? s=None, int[1] dim=[-2,-1], str? norm=None) -> Tensor
```
This is the only reason I've marked this as draft. Technically, you can use `np.fft.fft2(x, axes=None)` as a weird alternative to `np.fft.fftn(x)`. Supporting this would require more changes to the code generation to support optional arrays with defaults other than `None`. 

This isn't really the intended way to use these functions though, so it might be okay to break compatibility here.

peterbell10 replied:
As per our discussion, this is fine since we're raising an error instead of giving the wrong answer compared to NumPy.

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -510,6 +510,119 @@ def test_fftn_invalid(self, device):
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
+    @onlyOnCPUAndCUDA
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
```
We now require NumPy in our test suite and don't need these checks anymore

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -510,6 +510,119 @@ def test_fftn_invalid(self, device):
+
+                # Once with dim defaulted
+                input_np = input.cpu().numpy()
+                expected = numpy_fn(input_np, s, axes=(-2, -1), norm=norm)
```
This is interesting. Should this not explicitly provide a value for axes to be consistent with the actual computation below?

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -35,6 +35,36 @@ inline Tensor ifft(const Tensor& self,
   return torch::fft_ifft(self, n, dim, norm);
 }
 
+/// Computes the 2 dimensional fast Fourier transform over given dimensions.
```
"over given" -> "over the given"

mruberry replied:
"2 dimensional" -> "2-dimensional"

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -35,6 +35,36 @@ inline Tensor ifft(const Tensor& self,
+  return torch::fft_fft2(self, s, dim, norm);
+}
+
+/// Computes the 2 dimensional fast Fourier transform over given dimensions.
```
Same change here.

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -99,6 +129,36 @@ inline Tensor irfft(const Tensor& self,
   return torch::fft_irfft(self, n, dim, norm);
 }
 
+/// Computes the 2 dimensional FFT of real input with onesided Hermitian output.
```
Should this be: "of a real input. Returns a onesided Hermitian output."?

mruberry commented on torch/csrc/api/include/torch/fft.h:
````diff
@@ -99,6 +129,36 @@ inline Tensor irfft(const Tensor& self,
+/// Example:
+/// ```
+/// auto t = torch::randn({128, 128}, dtype=kComplexDouble);
+/// torch::fft::irfftn(t);
````
This should be a call to irfft2, not irfftn

mruberry commented on torch/fft/__init__.py:
```diff
@@ -93,6 +93,110 @@
+fft2 = _add_docstr(_fft.fft_fft2, r"""
+fft2(input, s=None, dim=(-2, -1), norm=None) -> Tensor
+
+Computes the 2 dimensional discrete Fourier transform of :attr:`input`.
```
"2 dimensional" -> "2-dimensional"

mruberry commented on torch/fft/__init__.py:
```diff
@@ -93,6 +93,110 @@
+Computes the 2 dimensional discrete Fourier transform of :attr:`input`.
+
+Note:
+    :func:`~torch.fft.fft2` is a convenience alias for :func:`~torch.fft.fftn`
```
Instead of making this a note, how about adding a second sentence to the preamble like: "Equivalent to ..." and linking fftn and showing how to map the arguments from this to fftn?

peterbell10 replied:
@mruberry PTAL, I've changed it but am not entirely sure about "map[ping] the arguments from this to fftn". The arguments should be exactly the same, other than `dim` not accepting `None`.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -93,6 +93,110 @@
+    with new defaults for :attr:`dim`. There are no other implications (e.g. to
+    performance) of using one or the other.
+
+Note:
```
Should there be more documentation for this function or a third sentence that simply says, "See fftn."?

peterbell10 replied:
In this case the note is more specific by making it 2d-only. It also directs the user to `rfft2` instead of `rfftn`.

mruberry replied:
OK, I can buy that.

mruberry commented:
Hey @peterbell10!

Great to have these last few functions. As usual I've asked a few questions about the docs. Sorry about the review delay but review responsiveness should finally be back to normal. Looking forward to hearing your thoughts!

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

mruberry commented on torch/fft/__init__.py:
```diff
@@ -93,6 +93,102 @@
+""")
+
+ifft2 = _add_docstr(_fft.fft_ifft2, r"""
+ifftn(input, s=None, dim=(-2, -1), norm=None) -> Tensor
```
ifftn->ifft2

mruberry reviewed (approved):
Another awesome PR, @peterbell10! Just one minor correction in the docs. Ping me when you're ready.

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]

peterbell10 commented:
Fixed. Thanks @mruberry.

peterbell10 committed:
Update on "torch.fft: Two dimensional FFT functions"

This PR implements `fft2`, `ifft2`, `rfft2` and `irfft2`. These are the last functions required for `torch.fft` to match `numpy.fft`. If you look at either NumPy or SciPy you'll see that the 2-dimensional variants are identical to `*fftn` in every way, except for the default value of `axes`. In fact you can even use `fft2` to do general n-dimensional transforms.




[ghstack-poisoned]
